File: android/app/src/main/java/com/anonymous/MocapExpo/pose/PoseEngineModule.kt
package com.anonymous.MocapExpo.pose

import android.Manifest
import android.app.Activity
import android.content.pm.PackageManager
import androidx.camera.core.CameraSelector
import androidx.core.content.ContextCompat
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import java.util.concurrent.Executors

class PoseEngineModule(
  reactContext: ReactApplicationContext,
) : ReactContextBaseJavaModule(reactContext) {

  private companion object {
    const val MODULE_NAME = "PoseEngineModule"
    const val EVENT_STATUS = "PoseEngineStatus"

    const val ERROR_CAMERA_PERMISSION = "E_CAMERA_PERMISSION"
    const val ERROR_START = "E_START"
    const val ERROR_STOP = "E_STOP"
    const val ERROR_OPTIONS = "E_OPTIONS"
    const val ERROR_VIDEO_RECORDING = "E_VIDEO_RECORDING"
  }

  private enum class EngineState(val value: String) {
    IDLE("idle"),
    STARTING("starting"),
    RUNNING("running"),
    STOPPING("stopping"),
    ERROR("error"),
  }

  private val moduleExecutor = Executors.newSingleThreadExecutor()
  private val stateLock = Any()

  private var state = EngineState.IDLE
  private var hasListeners = false
  private var listenerCount = 0
  private var lastTargetFps = 30

  override fun getName(): String = MODULE_NAME

  @ReactMethod
  fun addListener(_eventName: String) {
    moduleExecutor.execute {
      val shouldEmit = synchronized(stateLock) {
        listenerCount += 1
        val becameActive = !hasListeners
        hasListeners = true
        becameActive
      }
      if (shouldEmit) {
        sendStatus("listener_on")
      }
    }
  }

  @ReactMethod
  fun removeListeners(count: Int) {
    moduleExecutor.execute {
      synchronized(stateLock) {
        listenerCount = maxOf(0, listenerCount - count)
        hasListeners = listenerCount > 0
      }
    }
  }

  @ReactMethod
  fun ping(promise: Promise) {
    val result = Arguments.createMap()
    result.putBoolean("ok", true)
    result.putString("version", "camera-android-wham-upload-v1")
    result.putString("pipeline", "wham_video_upload")
    promise.resolve(result)
  }

  @ReactMethod
  fun setPreviewActive(active: Boolean, promise: Promise) {
    moduleExecutor.execute {
      if (!active) {
        synchronized(stateLock) {
          state = EngineState.STOPPING
        }
        sendStatus("stopping")
        PoseCameraSession.stop {
          synchronized(stateLock) {
            state = EngineState.IDLE
          }
          sendStatus("idle")
          promise.resolve(null)
        }
        return@execute
      }

      startCamera(lastTargetFps, promise)
    }
  }

  @ReactMethod
  fun start(options: ReadableMap, promise: Promise) {
    moduleExecutor.execute {
      lastTargetFps = maxOf(1, options.getOptionalInt("targetFps") ?: lastTargetFps)
      startCamera(lastTargetFps, promise)
    }
  }

  @ReactMethod
  fun stop(promise: Promise) {
    moduleExecutor.execute {
      synchronized(stateLock) {
        state = EngineState.STOPPING
      }
      sendStatus("stopping")
      PoseCameraSession.stop {
        synchronized(stateLock) {
          state = EngineState.IDLE
        }
        sendStatus("idle")
        promise.resolve(null)
      }
    }
  }

  @ReactMethod
  fun startVideoRecording(options: ReadableMap, promise: Promise) {
    moduleExecutor.execute {
      val takeId = options.getOptionalString("takeId")
      if (takeId.isNullOrBlank()) {
        promise.reject(ERROR_OPTIONS, "takeId is required")
        return@execute
      }

      val fps = maxOf(1, options.getOptionalInt("fps") ?: lastTargetFps)
      lastTargetFps = fps
      val activity = currentActivityOrNull()
      if (activity == null) {
        promise.reject(ERROR_START, "Video recording failed: no active activity.")
        return@execute
      }

      if (!hasCameraPermission()) {
        promise.reject(ERROR_CAMERA_PERMISSION, "Camera permission denied")
        return@execute
      }

      PoseCameraSession.startRecording(
        reactApplicationContext,
        activity,
        PoseCameraSession.RecordingOptions(
          takeId = takeId,
          fps = fps,
        ),
      ) { error ->
        if (error != null) {
          sendStatus("video_recording_error", mapOf("message" to (error.message ?: error.toString())))
          promise.reject(
            ERROR_VIDEO_RECORDING,
            "Video recording failed to start: ${error.message}",
            error,
          )
        } else {
          synchronized(stateLock) {
            state = EngineState.RUNNING
          }
          sendStatus("video_recording_started")
          promise.resolve(null)
        }
      }
    }
  }

  @ReactMethod
  fun stopVideoRecording(promise: Promise) {
    moduleExecutor.execute {
      PoseCameraSession.stopRecording { result, error ->
        if (error != null) {
          sendStatus("video_recording_error", mapOf("message" to (error.message ?: error.toString())))
          promise.reject(
            ERROR_VIDEO_RECORDING,
            "Video recording failed to stop: ${error.message}",
            error,
          )
          return@stopRecording
        }

        if (result == null) {
          promise.reject(ERROR_VIDEO_RECORDING, "Video recording returned no result")
          return@stopRecording
        }

        sendStatus("video_recording_stopped")
        promise.resolve(result.toWritableMap())
      }
    }
  }

  override fun invalidate() {
    PoseCameraSession.stop()
    moduleExecutor.shutdownNow()
    super.invalidate()
  }

  private fun startCamera(targetFps: Int, promise: Promise) {
    if (!hasCameraPermission()) {
      synchronized(stateLock) {
        state = EngineState.ERROR
      }
      sendStatus("error_camera_permission_denied")
      promise.reject(ERROR_CAMERA_PERMISSION, "Camera permission denied")
      return
    }

    val activity = currentActivityOrNull()
    if (activity == null) {
      synchronized(stateLock) {
        state = EngineState.ERROR
      }
      sendStatus("error_start_failed", mapOf("message" to "No active activity."))
      promise.reject(ERROR_START, "Camera start failed: no active activity.")
      return
    }

    synchronized(stateLock) {
      state = EngineState.STARTING
    }
    sendStatus("starting", mapOf("targetFps" to targetFps))
    PoseCameraSession.start(
      reactApplicationContext,
      activity,
      PoseCameraSession.Config(
        lensFacing = CameraSelector.LENS_FACING_BACK,
        fps = targetFps,
      ),
      onFrame = null,
      onError = { message ->
        sendStatus("camera_error", mapOf("message" to message))
      },
    ) { error ->
      if (error != null) {
        synchronized(stateLock) {
          state = EngineState.ERROR
        }
        sendStatus("error_start_failed", mapOf("message" to (error.message ?: "Unknown error")))
        promise.reject(ERROR_START, "Camera start failed: ${error.message}", error)
        return@start
      }

      synchronized(stateLock) {
        state = EngineState.RUNNING
      }
      sendStatus("running", mapOf("pipeline" to "wham_video_upload", "targetFps" to targetFps))
      promise.resolve(null)
    }
  }

  private fun hasCameraPermission(): Boolean {
    return ContextCompat.checkSelfPermission(
      reactApplicationContext,
      Manifest.permission.CAMERA,
    ) == PackageManager.PERMISSION_GRANTED
  }

  private fun currentActivityOrNull(): Activity? = reactApplicationContext.currentActivity

  private fun sendStatus(
    status: String,
    extra: Map<String, Any?> = emptyMap(),
  ) {
    val shouldEmit = synchronized(stateLock) { hasListeners }
    if (!shouldEmit) {
      return
    }

    val payload = Arguments.createMap()
    payload.putString("status", status)
    payload.putString("engineState", synchronized(stateLock) { state.value })

    extra["message"]?.let { payload.putString("message", it as String) }
    extra["pipeline"]?.let { payload.putString("pipeline", it as String) }
    extra["targetFps"]?.let { payload.putInt("targetFps", it as Int) }

    reactApplicationContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit(EVENT_STATUS, payload)
  }

  private fun ReadableMap.getOptionalString(key: String): String? =
    if (hasKey(key) && !isNull(key)) getString(key) else null

  private fun ReadableMap.getOptionalInt(key: String): Int? =
    if (hasKey(key) && !isNull(key)) getInt(key) else null
}

File: ios/MocapExpo/pose/PoseEngineModule.swift
import AVFoundation
import Foundation
import React
import Vision

@objc(PoseEngineModule)
final class PoseEngineModule: RCTEventEmitter {
    private enum EngineState: String {
        case idle
        case starting
        case running
        case stopping
        case error
    }

    private enum ErrorCode {
        static let cameraPermission = "E_CAMERA_PERMISSION"
        static let start = "E_START"
        static let stop = "E_STOP"
        static let options = "E_OPTIONS"
        static let recording = "E_VIDEO_RECORDING"
    }

    private static let eventStatus = "PoseEngineStatus"
    private static let eventFrame = "PoseEngineFrame"

    private let moduleQueue = DispatchQueue(label: "com.mocapexpo.camera.module")
    private let stateLock = NSLock()

    private var engineState = EngineState.idle
    private var hasListeners = false
    private var previewActive = false
    private var lastTargetFps = 30
    private var poseFrameId = 0
    private var lastPoseFrameSentAtMs = 0.0
    private let poseFrameIntervalMs = 1000.0 / 15.0

    override static func requiresMainQueueSetup() -> Bool {
        false
    }

    override func supportedEvents() -> [String]! {
        [Self.eventStatus, Self.eventFrame]
    }

    override func startObserving() {
        stateLock.lock()
        hasListeners = true
        stateLock.unlock()
        sendStatus("listener_on")
    }

    override func stopObserving() {
        stateLock.lock()
        hasListeners = false
        stateLock.unlock()
    }

    @objc
    func ping(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        resolve([
            "ok": true,
            "version": "camera-ios-wham-upload-v1",
            "pipeline": "wham_video_upload"
        ])
    }

    @objc
    func setPreviewActive(
        _ active: Bool,
        resolver resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        moduleQueue.async {
            self.previewActive = active

            if !active {
                PoseCameraSession.shared.stop {
                    self.setEngineState(.idle)
                    self.sendStatus("idle")
                    resolve(nil)
                }
                return
            }

            guard self.hasCameraPermission() else {
                self.setEngineState(.error)
                self.sendStatus("error_camera_permission_denied")
                reject(ErrorCode.cameraPermission, "Camera permission denied", nil)
                return
            }

            self.setEngineState(.starting)
            self.sendStatus("starting")
            PoseCameraSession.shared.start(
                fps: self.lastTargetFps,
                onFrame: self.poseFrameHandler(),
                onError: { [weak self] message in
                    self?.sendStatus("camera_error", extra: ["message": message])
                }
            ) { error in
                if let error {
                    self.setEngineState(.error)
                    self.sendStatus("error_start_failed", extra: ["message": error.localizedDescription])
                    reject(
                        ErrorCode.start,
                        "Preview start failed: \(error.localizedDescription)",
                        error
                    )
                } else {
                    self.setEngineState(.running)
                    self.sendStatus("running", extra: ["pipeline": "wham_video_upload"])
                    resolve(nil)
                }
            }
        }
    }

    @objc
    func start(
        _ options: NSDictionary,
        resolver resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        moduleQueue.async {
            self.lastTargetFps = max(1, options.intValue("targetFps") ?? self.lastTargetFps)
            self.setPreviewActive(true, resolver: resolve, rejecter: reject)
        }
    }

    @objc
    func stop(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        moduleQueue.async {
            self.setEngineState(.stopping)
            self.sendStatus("stopping")
            PoseCameraSession.shared.stop {
                self.setEngineState(.idle)
                self.sendStatus("idle")
                resolve(nil)
            }
        }
    }

    @objc
    func startVideoRecording(
        _ options: NSDictionary,
        resolver resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        moduleQueue.async {
            guard let takeId = options.stringValue("takeId"), !takeId.isEmpty else {
                reject(ErrorCode.options, "takeId is required", nil)
                return
            }

            let fps = max(1, options.intValue("fps") ?? self.lastTargetFps)
            let orientation = options.stringValue("orientation") ?? "portrait"
            self.lastTargetFps = fps

            let startRecording = {
                PoseCameraSession.shared.startRecording(
                    takeId: takeId,
                    fps: fps,
                    orientation: orientation
                ) { [weak self] error in
                    guard let self else { return }
                    if let error {
                        self.sendStatus(
                            "video_recording_error",
                            extra: ["message": error.localizedDescription]
                        )
                        reject(
                            ErrorCode.recording,
                            "Video recording failed to start: \(error.localizedDescription)",
                            error
                        )
                        return
                    }

                    self.sendStatus("video_recording_started")
                    resolve(nil)
                }
            }

            PoseCameraSession.shared.start(
                fps: fps,
                onFrame: self.poseFrameHandler(),
                onError: { [weak self] message in
                    self?.sendStatus("camera_error", extra: ["message": message])
                }
            ) { error in
                if let error {
                    self.sendStatus("video_recording_error", extra: ["message": error.localizedDescription])
                    reject(
                        ErrorCode.recording,
                        "Camera failed to start before recording: \(error.localizedDescription)",
                        error
                    )
                    return
                }
                self.setEngineState(.running)
                startRecording()
            }
        }
    }

    @objc
    func stopVideoRecording(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        moduleQueue.async {
            PoseCameraSession.shared.stopRecording { [weak self] result, error in
                guard let self else { return }
                if let error {
                    self.sendStatus(
                        "video_recording_error",
                        extra: ["message": error.localizedDescription]
                    )
                    reject(
                        ErrorCode.recording,
                        "Video recording failed to stop: \(error.localizedDescription)",
                        error
                    )
                    return
                }

                guard let result else {
                    reject(ErrorCode.recording, "Video recording returned no result", nil)
                    return
                }

                self.sendStatus("video_recording_stopped")
                resolve(result.asDictionary())
            }
        }
    }

    override func invalidate() {
        PoseCameraSession.shared.stop {}
        super.invalidate()
    }

    private func setEngineState(_ state: EngineState) {
        stateLock.lock()
        engineState = state
        stateLock.unlock()
    }

    private func hasCameraPermission() -> Bool {
        AVCaptureDevice.authorizationStatus(for: .video) == .authorized
    }

    private func sendStatus(_ status: String, extra: [String: Any?] = [:]) {
        stateLock.lock()
        let shouldEmit = hasListeners
        let stateValue = engineState.rawValue
        stateLock.unlock()

        guard shouldEmit else { return }

        var payload: [String: Any] = [
            "status": status,
            "engineState": stateValue
        ]

        for (key, value) in extra {
            if let value {
                payload[key] = value
            }
        }

        DispatchQueue.main.async { [weak self] in
            self?.sendEvent(withName: Self.eventStatus, body: payload)
        }
    }

    private func poseFrameHandler() -> (CMSampleBuffer) -> Void {
        { [weak self] sampleBuffer in
            self?.handlePoseFrame(sampleBuffer)
        }
    }

    private func handlePoseFrame(_ sampleBuffer: CMSampleBuffer) {
        let nowMs = Date().timeIntervalSince1970 * 1000.0

        stateLock.lock()
        let shouldAnalyze = hasListeners && nowMs - lastPoseFrameSentAtMs >= poseFrameIntervalMs
        if shouldAnalyze {
            lastPoseFrameSentAtMs = nowMs
        }
        stateLock.unlock()

        guard shouldAnalyze else { return }

        guard let imageBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else { return }
        let pixelWidth = CVPixelBufferGetWidth(imageBuffer)
        let pixelHeight = CVPixelBufferGetHeight(imageBuffer)
        let imageOrientation = visionOrientation(
            pixelWidth: pixelWidth,
            pixelHeight: pixelHeight
        )
        let normalizedSize = normalizedImageSize(
            pixelWidth: pixelWidth,
            pixelHeight: pixelHeight,
            orientation: imageOrientation
        )

        let request = VNDetectHumanBodyPoseRequest()
        let handler = VNImageRequestHandler(
            cmSampleBuffer: sampleBuffer,
            orientation: imageOrientation,
            options: [:]
        )

        do {
            try handler.perform([request])
            guard let observation = request.results?.first else { return }
            let points = try observation.recognizedPoints(.all)
            let landmarks = mediapipeLandmarks(from: points)
            guard landmarks.contains(where: { $0 > 0 }) else { return }

            stateLock.lock()
            poseFrameId += 1
            let frameId = poseFrameId
            stateLock.unlock()

            let payload: [String: Any] = [
                "ts": nowMs,
                "frameId": frameId,
                "fps": 15,
                "trackingProfile": "pose",
                "requestedTrackingProfile": "pose",
                "sourceDevice": "ios",
                "coordinateSpace": "image_normalized",
                "imageWidth": normalizedSize.width,
                "imageHeight": normalizedSize.height,
                "inputImageWidth": pixelWidth,
                "inputImageHeight": pixelHeight,
                "videoOrientation": "portrait",
                "cameraPosition": "back",
                "isMirrored": false,
                "orientationCorrection": String(describing: imageOrientation),
                "landmarks": landmarks
            ]

            DispatchQueue.main.async { [weak self] in
                self?.sendEvent(withName: Self.eventFrame, body: payload)
            }
        } catch {
            sendStatus("pose_frame_error", extra: ["message": error.localizedDescription])
        }
    }

    private func visionOrientation(
        pixelWidth: Int,
        pixelHeight: Int
    ) -> CGImagePropertyOrientation {
        if pixelHeight >= pixelWidth {
            return .up
        }
        return .right
    }

    private func normalizedImageSize(
        pixelWidth: Int,
        pixelHeight: Int,
        orientation: CGImagePropertyOrientation
    ) -> (width: Int, height: Int) {
        switch orientation {
        case .left, .leftMirrored, .right, .rightMirrored:
            return (width: pixelHeight, height: pixelWidth)
        default:
            return (width: pixelWidth, height: pixelHeight)
        }
    }

    private func mediapipeLandmarks(
        from points: [VNHumanBodyPoseObservation.JointName: VNRecognizedPoint]
    ) -> [Double] {
        var landmarks = Array(repeating: 0.0, count: 33 * 4)

        func write(_ index: Int, _ joint: VNHumanBodyPoseObservation.JointName) {
            guard let point = points[joint], point.confidence > 0.05 else { return }
            let offset = index * 4
            landmarks[offset] = Double(point.location.x)
            landmarks[offset + 1] = Double(1.0 - point.location.y)
            landmarks[offset + 2] = 0.0
            landmarks[offset + 3] = Double(point.confidence)
        }

        write(0, .nose)
        write(2, .leftEye)
        write(5, .rightEye)
        write(7, .leftEar)
        write(8, .rightEar)
        write(11, .leftShoulder)
        write(12, .rightShoulder)
        write(13, .leftElbow)
        write(14, .rightElbow)
        write(15, .leftWrist)
        write(16, .rightWrist)
        write(23, .leftHip)
        write(24, .rightHip)
        write(25, .leftKnee)
        write(26, .rightKnee)
        write(27, .leftAnkle)
        write(28, .rightAnkle)

        return landmarks
    }
}

private extension NSDictionary {
    func stringValue(_ key: String) -> String? {
        guard let value = self[key], !(value is NSNull) else { return nil }
        return value as? String
    }

    func intValue(_ key: String) -> Int? {
        guard let value = self[key], !(value is NSNull) else { return nil }
        return (value as? NSNumber)?.intValue
    }
}
